magento/pwa-studio
# [bug]: Clears sensitive data on logout, login, and checkout

## TODO

Waiting for other checkout page PRs to get merged to verify that all necessary connection keys have been set.

## Description

We store data in several places - local storage (using persist cache AND browser persistence) and internal memory. When a user performs some action that indicates a change of user or a completion of some task (such as checkout) we should clear all data that is sensitive or no longer necessary. The purpose of this PR is to ensure that we clear/reset **all** of these possible data stores and not just one or the other. Since we use cachePersist we have to take some [manual steps](https://github.com/apollographql/apollo-cache-persist/issues/34#issuecomment-371177206).

Here's the events and the actions we should take:

Login:
* Clear stale cart (or merge cart?) - done, we delete cached cart on login.

Logout:
* Clear cart - done, we delete cached cart on logout.
* Clear user data - we use no-cache for user data outside of cart so this won't be necessary.

Checkout:
* Clear cart - TODO - waiting for #2288 

Gotchas:

Our use of `useAwaitQuery` does not handle replaying queries so we can't use `resetStore`. Our options thus are to either blow the entire thing away with `clearStore` or wipe specific cache items.

Since we're in here we may want to consider adding logic to [handle schema changes](https://github.com/apollographql/apollo-cache-persist#ive-had-a-breaking-schema-change-how-do-i-migrate-or-purge-my-cache). Would love thoughts on this.

## Related Issue





Closes [PWA-457](https://jira.corp.magento.com/browse/PWA-457).

## Acceptance 



### Verification Stakeholders

### Specification



### Verification Steps


1. See that there is now a single `Cart` entry within the apollo cache.
2. See that there are no longer cart ids used in keys within ROOT_QUERY or ROOT_MUTATION within apollo cache.
3. Regress all things cart and user.  An authed user should not see guest user info such as cart or other info from a previous checkout. A guest user should also not see data from a previously logged in user. To check this, log in and use mini cart to attach some data such as addresses/etc to the cart. Log out. After the page refreshes the old cart data should be gone from apollo memory (use apollo dev tools) and local storage. There should be a new cart id for the guest user. Add some more data to the new cart and log in. The cart you just added items/data to should be gone and the new cart should not show any old data (it'll probably show cart data from the initial cart interaction.

## Screenshots / Screen Captures (if appropriate)

## Checklist

* I have updated the documentation accordingly, if necessary.
* I have added tests to cover my changes, if necessary.
